### PR TITLE
Fix C++/CLI applications which use __declspec(appdomain)

### LINKED ISSUE #110365: C++/CLI __declspec(appdomain) static variables cause a C++/CLI binary to fail to load.
### Description

If a C++/CLI dll has a global variable that is marked with __declspec(appdomain) the binary will fail to load. It will produce an error like the following:

```
System.IO.FileLoadException: Assembly is still being loaded. (0x80131016 (MSEE_E_ASSEMBLYLOADINPROGRESS))
   at System.Reflection.RuntimeAssembly.InternalLoad(NativeAssemblyNameParts* pAssemblyNameParts, ObjectHandleOnStack requestingAssembly, StackCrawlMarkHandle stackMark, Boolean throwOnFileNotFound, ObjectHandleOnStack assemblyLoadContext, ObjectHandleOnStack retAssembly)
   at System.Reflection.RuntimeAssembly.InternalLoad(AssemblyName assemblyName, StackCrawlMark& stackMark, AssemblyLoadContext assemblyLoadContext, RuntimeAssembly requestingAssembly, Boolean throwOnFileNotFound)
   at System.Reflection.Assembly.Load(String assemblyString)
   at NativeCallingManaged.NativeCallingManaged.TestEntryPoint() in runtime\src\tests\Interop\IJW\NativeCallingManaged\NativeCallingManaged.cs:line 25
   at __GeneratedMainWrapper.Main() in runtime\artifacts\tests\coreclr\obj\windows.x64.Debug\Managed\Interop\IJW\NativeCallingManaged\NativeCallingManaged\XUnitWrapperGenerator\XUnitWrapperGenerator.XUnitWrapperGenerator\SimpleRunner.g.cs:line 7
```

### Reproduction Steps

The bug report was provided privately to the CoreCLR team, and a repro will be provided as part of the fix.

### Expected behavior

Assembly should be loadable.

### Actual behavior

```
System.IO.FileLoadException: Assembly is still being loaded. (0x80131016 (MSEE_E_ASSEMBLYLOADINPROGRESS))
```

### Regression?

This is a regression in .NET 9 from .NET 8 related to the work to rebuild how static variables are handled.

### Known Workarounds

Remove `__declspec(appdomain)` from the definition of the global variable. On .NET 9 this has very little semantic effect (it can change initialization order relative to other global variables), causes a minor slowdown when using the variable, and triggers this bug.

### Configuration

This only reproduces for C++/CLI binaries as they have a unique startup path which causes the `<module>` type to be loaded before it would normally be permitted.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by davidwrighton**
Do not eagerly allocate static variable space while loading the assembly in use. This avoids possible recursive loading issues found in C++/CLI codebases. Repurpose the NativeCallingManaged test to subsume this particular regression test case.

Fix #110365

### CODE REVIEW

File: src/coreclr/vm/methodtable.cpp [hunk 1]
```diff
@@ -4317,7 +4317,10 @@ void MethodTable::DoFullyLoad(Generics::RecursionGraph * const pVisited,  const
         ClassLoader::ValidateMethodsWithCovariantReturnTypes(this);
     }
 
-    if ((level == CLASS_LOADED) && CORDisableJITOptimizations(this->GetModule()->GetDebuggerInfoBits()) && !HasInstantiation())
+    if ((level == CLASS_LOADED) && 
+        CORDisableJITOptimizations(this->GetModule()->GetDebuggerInfoBits()) &&
+        !HasInstantiation() &&
+        !GetModule()->GetAssembly()->IsLoading()) // Do not do this during the vtable fixup stage of C++/CLI assembly loading. See https://github.com/dotnet/runtime/issues/110365
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/methodtable.cpp:4323]: The comment is most appreciated. Two questions:

1) How do we know the "vtable fixup stage" is occurring? Is that `CLASS_LOADED`?
2) It seems like `IsLoading()` is implying that the type is being loaded _during_ assembly load (i.e., global statics are being initialized at that time?) and that combination is causing stress. This begs the question, is this limited to C++/CLI scenarios where statics for them are initialized at a time outside the ECMA loading process?

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Repurposing this test makes sense to me. I have a question on the type loading sequence though.

File: src/coreclr/vm/methodtable.cpp (hunk 1 above)
Comment (davidwrighton) [reply src/coreclr/vm/methodtable.cpp:4323]: Yes, these things are being loaded during assembly load. Technically this change will also impact scenarios where we have a module constructor, but really that's not a big deal either. This eager static's creation is really only here to make the debugger experience a bit prettier, so deferring that case too should be ok. For the loading stage.... this happens while the assembly load level is `FILE_LOAD_DELIVER_EVENTS` and is transitioning to `FILE_LOAD_VTABLE_FIXUPS`, but really we can run into issues any time the m_level is less than `FILE_LOADED`  This fix is a little aggresive as it doesn't just check for `FILE_LOADED` it checks for if its currently loading the assembly at all, but a slightly bigger fix is preferable here, as the diagnostic consequences of not filling this in are minimal, and the failure mode is so hard to debug.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**